include/nil/crypto3/zk/snark/systems/plonk/redshift/verifier.hpp

```diff
@@ -0,0 +1,160 @@
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in all
+// copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+// SOFTWARE.
+//---------------------------------------------------------------------------//
+
+#ifndef CRYPTO3_ZK_PLONK_REDSHIFT_VERIFIER_HPP
+#define CRYPTO3_ZK_PLONK_REDSHIFT_VERIFIER_HPP
+
+#include <nil/crypto3/math/polynomial/polynomial.hpp>
+
+#include <nil/crypto3/zk/snark/commitments/fri_commitment.hpp>
+#include <nil/crypto3/zk/snark/relations/constraint_satisfaction_problems/r1cs.hpp>
+#include <nil/crypto3/zk/snark/systems/plonk/redshift/types.hpp>
+
+namespace nil {
+    namespace crypto3 {
+        namespace zk {
+            namespace snark {
+
+                template<typename FieldType, std::size_t lambda, std::size_t m = 2>
+                class redshift_verifier {
+
+                    using types_policy = detail::redshift_types_policy<FieldType>;
+                    using transcript_manifest = types_policy::prover_fiat_shamir_heuristic_manifest<6>;
+
+                    typedef hashes::sha2<256> merkle_hash_type;
+                    typedef hashes::sha2<256> transcript_hash_type;
+
+                    constexpr static const std::size_t k = ...;
+                    constexpr static const std::size_t r = ...;
+
+                    constexpr static const typename FieldType::value_type omega =
+                        algebra::get_root_of_unity<FieldType>() typedef list_polynomial_commitment_scheme<
+                            FieldType, merkle_hash_type, lambda, k, r, m>
+                            lpc;
+
+                public:
+                    static inline bool process(const types_policy::verification_key_type &verification_key,
+                                               const types_policy::primary_input_type &primary_input,
+                                               const types_policy::proof_type<lpc> &proof) {
+
+                        std::size_t N_wires = ...;
+                        std::size_t N_perm = ...;
+                        std::size_t N_sel = ...;
+                        std::size_t N_const = ...;
+
+                        fiat_shamir_heuristic<transcript_manifest, hashes::sha2> transcript;
+
+                        ... setup_values = ...;
+                        transcript(setup_values);
```

Comment:
This seems to be not defined by the spec.

Reply:
Fixed.